tap_intercom/stream.py

```diff
@@ -21,40 +21,63 @@ def do_sync(self, tap_stream_id: str, state: Optional[datetime] = None):
             state=state, tap_stream_id=tap_stream_id
         )
         with singer.metrics.record_counter(tap_stream_id) as counter:
+            # The Intercom API doesn't allow for iterating through more than 10k companies
+            # using the list endpoint with pagination, instead requiring us to use their
+            # special "scroll" endpoint.
+            if tap_stream_id == "companies":
+                data = self.intercom.scroll_companies()
+                # Since we're guarding against records updated before the start date, it's
+                # safe to assume that the latest updated record in the end will be equal to
+                # or greater than the start date.
+                latest_updated_record = start_date
 
-            try:
-                data = self.intercom.get_records(tap_stream_id)
                 for record, replication_value in data:
-
-                    if replication_value and (
-                        start_date >= replication_value or end_date <= replication_value
-                    ):
+                    if replication_value < start_date:
                         continue
 
                     singer.write_record(tap_stream_id, record)
                     counter.increment(1)
-                    if not replication_value:
-                        continue
 
-                    new_bookmark = replication_value
-                    if not prev_bookmark:
+                    latest_updated_record = max(
+                        latest_updated_record, replication_value
+                    )
+
+                self.__advance_bookmark(state, latest_updated_record, tap_stream_id)
```

Comment:
I see you write the bookmark in the end. If the tap break in the middle of syncing company data, there will be no bookmark and next time we need to sync all over again.